[Infrastructure UI] Hosts table sorting and pagination

jennypavlova opened the pull request:
Closes https://github.com/elastic/kibana/issues/143510
## Summary

This PR adds pagination and column sorting to the hosts' table. I use default pagination settings (10 per page with the options to show 25 and 50). In order to avoid a change in the snapshot API for now I implemented it the same way as the inventory table view. If we see that the loading is slow and we need to optimize it we can change the snapshot API as well   
<img width="1448" alt="image" src="https://user-images.githubusercontent.com/14139027/196728634-b49a3abe-a5b7-4a03-8209-3e92ace8f0d9.png">

Testing:
I opened the inventory page in table view and compared the way pagination and sorting work there with the hosts' table using different sorting types ( ascending/descending order for each column ) 

Something I found out with the sorting while testing: (using `EuiInMemoryTable` ) When we have a text value (like the host.name) and we sort it (alphabetically for example) we will show first all the hosts starting with a capital letter (sorted alphabetically) and then the other hosts (sorted alphabetically) so the order will be for example **H**osta, Hostb, **a**host, bhost. It works the same way on the inventory page so I guess the sorting is implemented to be case-sensitive on purpose.

Old PR: https://github.com/neptunian/kibana/pull/3

elasticmachine commented:
Pinging @elastic/infra-monitoring-ui (Team:Infra Monitoring UI)

jennypavlova commented:
@crespocarlos Based on our discussion on the previous [PR](https://github.com/neptunian/kibana/pull/3) I am opening this one against `main` ( I got a warning that the comments may be lost if I change the old one so I will keep it to track the comments there until we merge this one)

crespocarlos commented on x-pack/plugins/infra/public/pages/metrics/hosts/components/hosts_table_columns.tsx:
```diff
@@ -10,28 +10,29 @@ import React from 'react';
+  name: string;
 }
 
 export interface HostMetics {
```
I think we could achieve the same results, without changing the types here. I played with it and `EuiInMemoryTable` complains that `item` type is not compatible because `name` is now declared as a mandatory field. If we make it optional, it should work without these changes here and on `HostsTableColumns` - correct me if I'm wrong or missing something.

crespocarlos commented on x-pack/plugins/infra/public/pages/metrics/hosts/components/hosts_table_columns.tsx:
```diff
@@ -41,59 +42,63 @@ export const HostsTableColumns: Array<EuiBasicTableColumn<HostNodeRow>> = [
     }),
     field: 'os',
+    sortable: true,
     render: (os: string) => <EuiText size="s">{os ?? '-'}</EuiText>,
```
Maybe we don't need this `os ?? '-'` anymore because we're handling null values in `use_host_tables`?

jennypavlova replied:
yes, true, forgot it after the change. Removed 👍

crespocarlos commented on x-pack/plugins/infra/public/pages/metrics/hosts/hooks/use_host_table.ts:
```diff
@@ -7,19 +7,28 @@
-    return nodes.map(({ metrics, path }) => ({
-      ...last(path),
+  const items: HostNodeRow[] = useMemo(() => {
+    const valuesMapping: Record<keyof HostMetics, 'value' | 'avg' | 'max'> = {
```
This is a smart solution :). If we don't need to make the changes on `hosts_table_column` we may not need this anymore - which would be even simpler

jennypavlova replied:
I changed the columns because for the sorting we need the fields to have the value as a number or string - the value we want to use for sorting the table - so for example `memory: {avg: 1, value: 2, max: 3}` should be `memory: 1` because we show the `avg` in the table and for sorting the column the average value (`1`)  will be used ( similar to all the columns ). That's why I changed the structure there.

crespocarlos commented on x-pack/plugins/infra/public/pages/metrics/hosts/hooks/use_host_table.ts:
```diff
@@ -7,19 +7,28 @@
+    };
+    return nodes.map(({ metrics, path, name }) => ({
+      name,
+      os: last(path)?.os ?? '-',
```
If you prefer, you can get rid of `lodash` by doing the following: 

```
const lastPath = { ...path.at(-1), os: path.at(-1)?.os ?? '-' };
```

new TS cool feature

crespocarlos commented on x-pack/plugins/infra/public/pages/metrics/hosts/hooks/use_host_table.ts:
```diff
@@ -7,19 +7,28 @@
       ...metrics.reduce((data, metric) => {
-        data[metric.name as keyof HostMetics] = metric;
+        const metricName = metric.name as keyof HostMetics;
+        data[metricName] = metric[valuesMapping[metricName]];
```
Should we hand;e null values here too? In case these any of these metrics are null, should we set the default value to `0`?

jennypavlova replied:
The `NumberCell` [component](https://github.com/elastic/kibana/blob/a59c0482ca87de4a65722f01cc2e713f2862ace1/x-pack/plugins/infra/public/components/infrastructure_node_metrics_tables/shared/components/number_cell.tsx#L18) will handle the `null` case and will display `N/A` in that case - I think it's more accurate when the metric is not present instead of showing `0` - what do you think?

crespocarlos reviewed (commented):
Cool! Paging and sorting works. Perhaps we could try to look into the type changes and see if we can try to simplify it.

jennypavlova commented:
@crespocarlos As discussed I changed the mapping so we can have the same metric values format as we have now and changed the field names inside `HostsTableColumns`. Also changed `label` to `name` and removed not used values from the `path` object.
